### PR TITLE
:bug: Reject rename if file already exists

### LINKED ISSUE #389: rename should be rejected if target filename exists
it will be overwritten if target filename exists now.

### PR DESCRIPTION
**Pull request by moldis**
Related to https://github.com/0chain/blobber/issues/389

### CODE REVIEW

File: code/go/0chain.net/blobbercore/allocation/renamefilechange.go [hunk 1]
```diff
@@ -29,6 +29,12 @@ func (rf *RenameFileChange) ProcessChange(ctx context.Context, change *Allocatio
 	if err != nil {
 		return nil, err
 	}
+
+	_, err = reference.GetObjectTree(ctx, rf.AllocationID, rf.NewName)
```
Comment (cnlangzi) [inline code/go/0chain.net/blobbercore/allocation/renamefilechange.go]: It is too heavy to load so many nodes for it?

```
	if path != "/" {
		db = db.Or("path LIKE ? AND allocation_id = ?", (path + "/%"), allocationID)
	} else {
		db = db.Or("path LIKE ? AND allocation_id = ?", (path + "%"), allocationID)
	}
```

do you think it is better to add a simple check function for better sql performance?

**Review by cnlangzi — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/allocation/renamefilechange.go (hunk 1 above)
Comment (moldis) [reply code/go/0chain.net/blobbercore/allocation/renamefilechange.go]: @cnlangzi Yes, the whole function is very messy and hard to read. I created separate issue to refactor it

File: code/go/0chain.net/blobbercore/allocation/renamefilechange.go [hunk 2]
```diff
@@ -25,10 +25,16 @@ func (rf *RenameFileChange) DeleteTempFile() error {
 }
 
 func (rf *RenameFileChange) ProcessChange(ctx context.Context, change *AllocationChange, allocationRoot string) (*reference.Ref, error) {
+	exists, _ := reference.PathExists(ctx, rf.AllocationID, rf.NewName)
+	if exists {
+		return nil, common.NewError("invalid_reference_path", "Invalid new name of path, it's already exists")
+	}
```
Comment (kushthedude) [inline code/go/0chain.net/blobbercore/allocation/renamefilechange.go]:
```suggestion
	isFilePresent, _ := reference.PathExists(ctx, rf.AllocationID, rf.NewName)
	if isFilePresent {
		return nil, common.NewError("invalid_reference_path", "file already exists")
	}
```

**Review by kushthedude — DISMISSED** [review]
Looks good to me @moldis, have made a small grammatical correction. Btw do you mind adding a small unit test to verify this fix?

File: code/go/0chain.net/blobbercore/reference/referencepath.go [hunk 3]
```diff
@@ -78,6 +78,30 @@ func GetReferencePathFromPaths(ctx context.Context, allocationID string, paths [
...
+
+	if path == "." || path == "/" {
+		return false, nil
+	}
+
+	var refs []Ref
+	db := datastore.GetStore().GetTransaction(ctx)
+
+	ref := Ref{Path: path, AllocationID: allocationID}
+	db = db.Where(ref).Or("path LIKE ? AND allocation_id = ?", (path + "/%"), allocationID)
```
Comment (cnlangzi) [inline code/go/0chain.net/blobbercore/reference/referencepath.go]: do you think `db = db.Where(ref)` is enough?

File: code/go/0chain.net/blobbercore/reference/referencepath.go [hunk 4]
```diff
@@ -78,6 +78,30 @@ func GetReferencePathFromPaths(ctx context.Context, allocationID string, paths [
...
+
+	err := db.Order("level, lookup_hash").Preload(ref.TableName(), func(tx *gorm.DB) *gorm.DB {
```
Comment (cnlangzi) [inline code/go/0chain.net/blobbercore/reference/referencepath.go]: `order by` and `preload` should be unnecessary here? please double check it. thanks.

**Review by cnlangzi — CHANGES_REQUESTED** [review]

**Review by cnlangzi — APPROVED** [review]

**Pull request merged**